### PR TITLE
New lexer: Lutin 

### PR DESCRIPTION
**Pull request by jahierwan**
@pyrmont

### CODE REVIEW

**Comment by pyrmont** [comment]
@jahierwan This still has the Lustre lexer files.

**Comment by jahierwan** [comment]
> @jahierwan This still has the Lustre lexer files.

oh, yes, I've forked my master copy which is under review, sorry. 
Is it ok if I wait #905 PR to be accepted and merged, so that I can merge this branch against master 
before pushing this branch? Or do you want me start from a fresh branch?

**Comment by pyrmont** [comment]
You won't need to merge it back into your master. Once it's accepted, it'll be squashed and merged into the official master. You'll just merge that into your master to bring yourself up to date.

**Comment by jahierwan** [comment]
ok, thanks.

**Comment by pyrmont** [comment]
@jahierwan I'm sorry—my git foo wasn't strong enough to work out how to unpick this branch so I basically nuked it and started over. There's now just the current commit with the files as they were after your most recent commit. I'll have a look and get back to you with any comments ASAP.

File: lib/rouge/lexers/lutin.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*- #
+#
```
Comment (pyrmont) [inline lib/rouge/lexers/lutin.rb]: This needs a frozen string literal pragma.

File: lib/rouge/lexers/lutin.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      def self.primitives
+        @primitives ||= Set.new %w(int real bool trace loop fby)
+      end
+
+      operator = %r([;,_!$%&*+./:<=>?@^|~#-]+)
+      id = /[a-z_][\w']*/i
+
+      state :root do
+        rule (/\s+/m), Text
+        rule (/false|true|[(][)]|\[\]/), Keyword::Constant
```
Comment (pyrmont) [inline lib/rouge/lexers/lutin.rb]: The parentheses and brackets shouldn't be part of this pattern.

File: lib/rouge/lexers/lutin.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        end
+
+        rule (/[(){}\[\];]+/), Punctuation
+        rule operator, Operator
+
+        rule (/-?\d[\d_]*(.[\d_]*)?(e[+-]?\d[\d_]*)/i), Num::Float
+        rule (/0x\h[\h_]*/i), Num::Hex
+        rule (/0o[0-7][0-7_]*/i), Num::Oct
+        rule (/0b[01][01_]*/i), Num::Bin
+        rule (/\d[\d_]*/), Num::Integer
```
Comment (pyrmont) [inline lib/rouge/lexers/lutin.rb]: The types of number here are more varied than in the Lustre lexer. Is that intentional?

**Comment by jahierwan** [comment]
> The Lutin lexer seems (at least superficially) almost identical to the Lustre lexer. You can use a lexer as the parent of another lexer and then simplify modify the parts that have changed (see [the C++ lexer](https://github.com/rouge-ruby/rouge/blob/9724e00e6aaeb5811d5df0b24bd163255c39b81c/lib/rouge/lexers/cpp.rb) for an example).
> 
> Are these languages related enough that changes to one will require changes to the other? Is one a superset of the other for example?

Clearly the syntax of Lutin tried to mimick as much as possible to the one of Lustre. Nevertheless, they are independent languages and modifying one does not imply some modification to the other

File: lib/rouge/lexers/lutin.rb (hunk 2 above)
Comment (jahierwan) [reply lib/rouge/lexers/lutin.rb]: ok

File: lib/rouge/lexers/lutin.rb (hunk 3 above)
Comment (jahierwan) [reply lib/rouge/lexers/lutin.rb]: Nope.

**Comment by pyrmont** [comment]
@jahierwan Thanks for working through this one. I thought it would be good to have the Lustre and Lutin lexers ready for the same release (I'm about to push that out now). Hope you find it useful!

**Comment by jahierwan** [comment]
Cool. Thanks for all the feedback.Cheers.R1--------- Original Message ---------
From: Michael Camilleri 
To: rouge-ruby/rouge 
Date: Tue Sep 03 19:08:12 GMT+02:00 2019
Subject: Re: [rouge-ruby/rouge] New lexer: Lutin  (#1307)

@jahierwan Thanks for working through this one. I thought it would be good to have the Lustre and Lutin lexers ready for the same release (I'm about to push that out now). Hope you find it useful!

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/rouge-ruby/rouge/pull/1307?email_source=notifications\u0026email_token=AAUIWL5T5BIXDLC3YSJBX23QH2KXZA5CNFSM4INB5NMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5Y4R2Q#issuecomment-527550698",
"url": "https://github.com/rouge-ruby/rouge/pull/1307?email_source=notifications\u0026email_token=AAUIWL5T5BIXDLC3YSJBX23QH2KXZA5CNFSM4INB5NMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5Y4R2Q#issuecomment-527550698",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Pull request merged**